### PR TITLE
Update Dockerfiles and CI

### PR DESCRIPTION
**Pull request by ruffsl**
Spinning out non-dumenten related CI changes from https://github.com/ros-planning/navigation2/pull/1731 into own seperate PR.

* Use build kit for persistent caching across docker builds
* Clean up dummy files for better scripting
* Simple ergonomic improvements and file clarity

### CODE REVIEW

File: .dockerhub/source.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+      -name "CATKIN_IGNORE" -o \
+      -name "COLCON_IGNORE" \
+      | xargs dirname | xargs rm -rf || true && \
+    colcon list --paths-only \
+      --packages-skip-up-to  \
+        $(cat $OVERLAY_WS/packages.txt | xargs) \
+      | xargs rm -rf
+
+# copy manifests for caching
+RUN mkdir -p /tmp/opt && \
```
Comment (SteveMacenski) [inline .dockerhub/source.Dockerfile:62]: What's the aim for caching the package.xml files?

File: Dockerfile [hunk 2]
```diff
@@ -91,14 +89,22 @@ RUN . $UNDERLAY_WS/install/setup.sh && \
...
 
 # source overlay from entrypoint
 ENV UNDERLAY_WS $UNDERLAY_WS
 ENV OVERLAY_WS $OVERLAY_WS
 RUN sed --in-place \
       's|^source .*|source "$OVERLAY_WS/install/setup.bash"|' \
       /ros_entrypoint.sh
+
+# test overlay build
+ARG RUN_TESTS
```
Comment (SteveMacenski) [inline Dockerfile:102]: Should we be doing tests as part of the build for the development docker image?

File: .dockerhub/source.Dockerfile (hunk 1 above)
Comment (ruffsl) [reply .dockerhub/source.Dockerfile:62]: In short this allows you to modify the local codebase in the repo without busting the docker build cache for cloning/building the underlay or re-installing package dependencies. These caches are selectively busted only if the .repos file or local package manifests are changed; and so this saves a load of dev time when build/testing locally using docker.

* https://github.com/ros-planning/navigation2/blob/a8df3f014ad9f1527086b79c1755cbb679897108/doc/continuous_integration/dockerfile.md#cacher-stage

This fancy multi staging is a workaround for the fact that docker's `COPY` directive does not have full globbing support, e.g. `COPY ./**/package.xml ./` does not work as expected. But still, more clever things can be done with multi staging.

* https://github.com/moby/moby/issues/39530#issuecomment-637905128

File: Dockerfile (hunk 2 above)
Comment (ruffsl) [reply Dockerfile:102]: I left the `ARG` empty to disable testing during the dockerfile build by default, as there is no need for DockerHub to run tests when CI is already running them nightly. It's more to provide users a quick means to run tests locally while compiling a fresh build, or to self document how to build and test the project's workspace.

File: .dockerhub/source.Dockerfile (hunk 1 above)
Comment (SteveMacenski) [reply .dockerhub/source.Dockerfile:62]: Isn't that bad if we don't break the cache for the overlay workspace only on new releases or a new pkg? There's lots of changes between formal releases

Comment (ruffsl) [reply .dockerhub/source.Dockerfile:62]: It's only for local dev builds, as builds on DockerHub are set to not use caching. If your building locally and haven't needed to change the package or repos file, and feel your docker build cache is getting stale, or find yourself with a speedy internet connection (some contributors don't have this at home), then you can use the `--no-cache` argument to docker build as mentioned in the Dockerfile to build from scratch.

**Comment by peci1** [comment]
I just saw the ROSCon lightning talk and wanted to have a look at the implementation. Great job digging through all the Docker docs and tutorials and distilling them into a thing like this ;) It's definitely an inspiration. I just wondered if I'm the only one who's lifting his eyebrows in front of the huge inefficiency of most Docker builds. I'm lucky I'm not alone =)

I also used the `mount=type=cache` technique in our Dockerfiles, and I found out it's also good to cache `/root/.cache/pip` because rosdep usually pulls some pip dependencies.

I'm not sure if it's really needed, but at some point I added `rw` to the cache definitions, like

    --mount=type=cache,target=/root/.cache/pip,rw

I was also experimenting with Git caching, but one has to be careful there to not break when people force-push or do some other silly stuff. My first idea was to cache bare git repos and utilize them using something like `git clone --reference`, but I couldn't find a way to integrate that with vcstool (or, frankly speaking, wstool we're still using). I ended up running wstool in a `mount=type=cache` directory and copying the files to workspace. I think that works as long as wstool/vcstool is able to update the cached state of repos to the new desired state. But I'm pretty sure there are cases where this would break.

**Comment by SteveMacenski** [comment]
It might be good for you to join our Slack group and talk to Ruffin or the rest of the team there ;-)

**Comment by peci1** [comment]
Not another Slack group, please :-D But if you feel these ideas could be leveraged, feel free to open an issue in this repo and invite me in. Or maybe rather in some tooling repo?

**Comment by ruffsl** [comment]
> I'm not sure if it's really needed, but at some point I added rw to the cache definitions, like

That looks new. What nuance effects does being explicit about the rewrite permissions to the cash volume have? By default, or when permissions are omitted, rewrite permissions are used, no? I may need to fix that then.

**Comment by peci1** [comment]
Ah, my bad, rw is for bind mounts. cache mounts have ro instead (so I assume rw is the default). https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/experimental.md#run---mounttypecache . When I was setting up our dockerfile, I had the impression that rw prevented some cache busts, but I can't tell for sure.

**Pull request merged**